File: benchmark/objc/TwitterTextBenchmark.m
// Copyright 2024 Robert Sayre
// Licensed under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0
//
// TwitterTextBenchmark.m
// Benchmark comparing old Obj-C twitter-text vs Rust FFI bindings

#import <Foundation/Foundation.h>

// Old implementation - use full path to avoid conflicts
#import "objc/lib/TwitterText.h"

// Rust FFI wrapper - aliased types
#import "rust/swift-bindings/objc/TwitterTextWrapper.h"

static const NSInteger ITERATIONS = 1000;
static const NSInteger WARMUP_ITERATIONS = 100;

// Rename old types to avoid conflicts with wrapper compatibility typedefs
typedef TwitterTextEntity OldTwitterTextEntity;

@interface BenchmarkData : NSObject
@property (nonatomic, strong) NSArray<NSDictionary *> *autolinkTests;
@property (nonatomic, strong) NSDictionary *extractTests;
@property (nonatomic, strong) NSDictionary *validateTests;
@property (nonatomic, strong) NSArray<NSDictionary *> *parseTests;
@end

@implementation BenchmarkData
@end

static NSDictionary *loadYAML(NSString *filename) {
    // YAML files are in benchmark/data/ directory
    NSString *path = [[NSBundle mainBundle] pathForResource:filename ofType:@"yml" inDirectory:@"benchmark/data"];
    if (!path) {
        // Try relative path for command-line execution
        path = [NSString stringWithFormat:@"benchmark/data/%@.yml", filename];
    }

    NSError *error = nil;
    NSString *content = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:&error];
    if (error) {
        NSLog(@"Error loading %@: %@", filename, error);
        return nil;
    }

    // Simple YAML parsing for our test data format
    // Since we control the format, we can use a simple parser
    NSMutableDictionary *result = [NSMutableDictionary dictionary];
    NSMutableDictionary *currentSection = nil;
    NSMutableArray *currentArray = nil;
    NSMutableDictionary *currentItem = nil;
    NSString *currentKey = nil;

    for (NSString *line in [content componentsSeparatedByString:@"\n"]) {
        NSString *trimmed = [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];

        if ([trimmed hasPrefix:@"#"] || trimmed.length == 0) continue;

        if ([line hasPrefix:@"tests:"]) {
            currentSection = [NSMutableDictionary dictionary];
            result[@"tests"] = currentSection;
        } else if ([line hasPrefix:@"  "] && ![line hasPrefix:@"    "]) {
            // Section like "  mentions:" or "  tweets:"
            NSString *sectionName = [[trimmed stringByReplacingOccurrencesOfString:@":" withString:@""] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            currentArray = [NSMutableArray array];
            if (currentSection) {
                currentSection[sectionName] = currentArray;
            } else {
                result[sectionName] = currentArray;
            }
        } else if ([trimmed hasPrefix:@"- "]) {
            currentItem = [NSMutableDictionary dictionary];
            [currentArray addObject:currentItem];
            // Parse inline key-value if present
            NSString *rest = [trimmed substringFromIndex:2];
            if ([rest containsString:@":"]) {
                NSRange colonRange = [rest rangeOfString:@":"];
                currentKey = [[rest substringToIndex:colonRange.location] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
                NSString *value = [[rest substringFromIndex:colonRange.location + 1] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
                if ([value hasPrefix:@"\""]) {
                    value = [value substringWithRange:NSMakeRange(1, value.length - 2)];
                }
                currentItem[currentKey] = value;
            }
        } else if ([trimmed containsString:@":"] && currentItem) {
            NSRange colonRange = [trimmed rangeOfString:@":"];
            currentKey = [[trimmed substringToIndex:colonRange.location] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            NSString *value = [[trimmed substringFromIndex:colonRange.location + 1] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            if ([value hasPrefix:@"\""]) {
                value = [value substringWithRange:NSMakeRange(1, value.length - 2)];
            }
            if (value.length > 0) {
                if ([value isEqualToString:@"true"]) {
                    currentItem[currentKey] = @YES;
                } else if ([value isEqualToString:@"false"]) {
                    currentItem[currentKey] = @NO;
                } else {
                    currentItem[currentKey] = value;
                }
            }
        }
    }

    return result;
}

static void printResults(NSString *operation, double oldOpsPerSec, double rustOpsPerSec) {
    double speedup = rustOpsPerSec / oldOpsPerSec;
    NSString *label = speedup >= 1 ? @"faster" : @"slower";
    double ratio = speedup >= 1 ? speedup : (1 / speedup);

    printf("\n%s (%ld iterations):\n", [operation UTF8String], (long)ITERATIONS);
    printf("  Old Obj-C: %.0f ops/sec\n", oldOpsPerSec);
    printf("  Rust FFI:  %.0f ops/sec\n", rustOpsPerSec);
    printf("  Result:    %.1fx %s\n", ratio, [label UTF8String]);
}

static void benchmarkExtract(NSDictionary *data) {
    NSDictionary *tests = data[@"tests"];
    NSArray *mentions = tests[@"mentions"];
    NSArray *urls = tests[@"urls"];
    NSArray *hashtags = tests[@"hashtags"];
    NSArray *cashtags = tests[@"cashtags"];

    // Warmup old
    for (NSInteger i = 0; i < WARMUP_ITERATIONS; i++) {
        for (NSDictionary *test in mentions) {
            [TwitterText mentionedScreenNamesInText:test[@"text"]];
        }
        for (NSDictionary *test in urls) {
            [TwitterText URLsInText:test[@"text"]];
        }
        for (NSDictionary *test in hashtags) {
            [TwitterText hashtagsInText:test[@"text"] checkingURLOverlap:YES];
        }
        for (NSDictionary *test in cashtags) {
            [TwitterText symbolsInText:test[@"text"] checkingURLOverlap:YES];
        }
    }

    // Benchmark old
    CFAbsoluteTime startOld = CFAbsoluteTimeGetCurrent();
    for (NSInteger i = 0; i < ITERATIONS; i++) {
        for (NSDictionary *test in mentions) {
            [TwitterText mentionedScreenNamesInText:test[@"text"]];
        }
        for (NSDictionary *test in urls) {
            [TwitterText URLsInText:test[@"text"]];
        }
        for (NSDictionary *test in hashtags) {
            [TwitterText hashtagsInText:test[@"text"] checkingURLOverlap:YES];
        }
        for (NSDictionary *test in cashtags) {
            [TwitterText symbolsInText:test[@"text"] checkingURLOverlap:YES];
        }
    }
    CFAbsoluteTime elapsedOld = CFAbsoluteTimeGetCurrent() - startOld;
    double oldOpsPerSec = ITERATIONS / elapsedOld;

    // Warmup rust (using TTTextEntity via wrapper)
    for (NSInteger i = 0; i < WARMUP_ITERATIONS; i++) {
        for (NSDictionary *test in mentions) {
            [TwitterText mentionedScreenNamesInText:test[@"text"]];
        }
        for (NSDictionary *test in urls) {
            [TwitterText URLsInText:test[@"text"]];
        }
        for (NSDictionary *test in hashtags) {
            [TwitterText hashtagsInText:test[@"text"] checkingURLOverlap:YES];
        }
        for (NSDictionary *test in cashtags) {
            [TwitterText symbolsInText:test[@"text"] checkingURLOverlap:YES];
        }
    }

    // Benchmark rust
    CFAbsoluteTime startRust = CFAbsoluteTimeGetCurrent();
    for (NSInteger i = 0; i < ITERATIONS; i++) {
        for (NSDictionary *test in mentions) {
            [TwitterText mentionedScreenNamesInText:test[@"text"]];
        }
        for (NSDictionary *test in urls) {
            [TwitterText URLsInText:test[@"text"]];
        }
        for (NSDictionary *test in hashtags) {
            [TwitterText hashtagsInText:test[@"text"] checkingURLOverlap:YES];
        }
        for (NSDictionary *test in cashtags) {
            [TwitterText symbolsInText:test[@"text"] checkingURLOverlap:YES];
        }
    }
    CFAbsoluteTime elapsedRust = CFAbsoluteTimeGetCurrent() - startRust;
    double rustOpsPerSec = ITERATIONS / elapsedRust;

    printResults(@"Extract", oldOpsPerSec, rustOpsPerSec);
}

static void benchmarkValidate(NSDictionary *data) {
    NSDictionary *tests = data[@"tests"];
    NSArray *tweets = tests[@"tweets"];

    // Warmup old - use tweetLength as proxy for validation
    for (NSInteger i = 0; i < WARMUP_ITERATIONS; i++) {
        for (NSDictionary *test in tweets) {
            [TwitterText tweetLength:test[@"text"]];
        }
    }

    // Benchmark old
    CFAbsoluteTime startOld = CFAbsoluteTimeGetCurrent();
    for (NSInteger i = 0; i < ITERATIONS; i++) {
        for (NSDictionary *test in tweets) {
            [TwitterText tweetLength:test[@"text"]];
        }
    }
    CFAbsoluteTime elapsedOld = CFAbsoluteTimeGetCurrent() - startOld;
    double oldOpsPerSec = ITERATIONS / elapsedOld;

    // Warmup rust
    for (NSInteger i = 0; i < WARMUP_ITERATIONS; i++) {
        for (NSDictionary *test in tweets) {
            [TwitterText tweetLength:test[@"text"]];
        }
    }

    // Benchmark rust
    CFAbsoluteTime startRust = CFAbsoluteTimeGetCurrent();
    for (NSInteger i = 0; i < ITERATIONS; i++) {
        for (NSDictionary *test in tweets) {
            [TwitterText tweetLength:test[@"text"]];
        }
    }
    CFAbsoluteTime elapsedRust = CFAbsoluteTimeGetCurrent() - startRust;
    double rustOpsPerSec = ITERATIONS / elapsedRust;

    printResults(@"Validate", oldOpsPerSec, rustOpsPerSec);
}

static void benchmarkParse(NSDictionary *data) {
    NSArray *tests = data[@"tests"];

    // Old implementation doesn't have TwitterTextParser, use tweetLength
    // Warmup old
    for (NSInteger i = 0; i < WARMUP_ITERATIONS; i++) {
        for (NSDictionary *test in tests) {
            [TwitterText tweetLength:test[@"text"]];
        }
    }

    // Benchmark old
    CFAbsoluteTime startOld = CFAbsoluteTimeGetCurrent();
    for (NSInteger i = 0; i < ITERATIONS; i++) {
        for (NSDictionary *test in tests) {
            [TwitterText tweetLength:test[@"text"]];
        }
    }
    CFAbsoluteTime elapsedOld = CFAbsoluteTimeGetCurrent() - startOld;
    double oldOpsPerSec = ITERATIONS / elapsedOld;

    // Rust FFI uses TTTextParser
    TTTextParser *parser = [TTTextParser defaultParser];

    // Warmup rust
    for (NSInteger i = 0; i < WARMUP_ITERATIONS; i++) {
        for (NSDictionary *test in tests) {
            [parser parseTweet:test[@"text"]];
        }
    }

    // Benchmark rust
    CFAbsoluteTime startRust = CFAbsoluteTimeGetCurrent();
    for (NSInteger i = 0; i < ITERATIONS; i++) {
        for (NSDictionary *test in tests) {
            [parser parseTweet:test[@"text"]];
        }
    }
    CFAbsoluteTime elapsedRust = CFAbsoluteTimeGetCurrent() - startRust;
    double rustOpsPerSec = ITERATIONS / elapsedRust;

    printResults(@"Parse Tweet", oldOpsPerSec, rustOpsPerSec);
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        printf("Twitter Text Benchmark: Old Obj-C vs Rust FFI\n");
        printf("==============================================\n");

        NSDictionary *extractData = loadYAML(@"extract");
        NSDictionary *validateData = loadYAML(@"validate");
        NSDictionary *parseData = loadYAML(@"parse");

        if (!extractData || !validateData || !parseData) {
            fprintf(stderr, "Failed to load benchmark data\n");
            return 1;
        }

        benchmarkExtract(extractData);
        benchmarkValidate(validateData);
        benchmarkParse(parseData);

        printf("\nDone.\n");
    }
    return 0;
}

File: benchmark/java/TwitterTextBenchmark.java
// Copyright 2025 Robert Sayre
// Licensed under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package benchmark;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import java.io.InputStream;
import java.util.List;
import java.util.Map;

/**
 * Benchmark comparing old Java twitter-text implementation vs Rust FFI bindings.
 * Uses result accumulation to thwart JIT escape analysis optimization.
 *
 * Note: This benchmark reflects realistic usage where Java code works with
 * Java Strings (UTF-16). The Rust FFI pays UTF-16 -> UTF-8 -> UTF-16 conversion
 * cost, which is inherent to the current binding design. A future optimization
 * could add byte[] APIs to avoid this overhead when input is already UTF-8.
 */
public class TwitterTextBenchmark {

    private static final int ITERATIONS = 1000;
    private static final int WARMUP_ITERATIONS = 100;

    // Global sink to prevent escape analysis from optimizing away results
    private static long sink = 0;

    // Old Java implementation
    private static final com.twitter.twittertext.Autolink OLD_AUTOLINK =
        new com.twitter.twittertext.Autolink(false);
    private static final com.twitter.twittertext.Extractor OLD_EXTRACTOR =
        new com.twitter.twittertext.Extractor();
    private static final com.twitter.twittertext.Validator OLD_VALIDATOR =
        new com.twitter.twittertext.Validator();

    public static void main(String[] args) throws Exception {
        System.out.println("Twitter Text Benchmark: Old Java vs Rust FFI");
        System.out.println("=============================================");

        // Load benchmark data
        ObjectMapper mapper = new ObjectMapper(new YAMLFactory());

        Map<String, Object> autolinkData = loadYaml(mapper, "autolink.yml");
        Map<String, Object> extractData = loadYaml(mapper, "extract.yml");
        Map<String, Object> validateData = loadYaml(mapper, "validate.yml");
        Map<String, Object> parseData = loadYaml(mapper, "parse.yml");

        benchmarkAutolink(autolinkData);
        benchmarkExtract(extractData);
        benchmarkValidateTweet(validateData);
        benchmarkValidateAll(validateData);
        benchmarkParse(parseData);

        // Print sink to ensure it's not optimized away
        System.out.println("\nDone. (checksum: " + sink + ")");
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> loadYaml(
        ObjectMapper mapper,
        String filename
    ) throws Exception {
        try (
            InputStream is = TwitterTextBenchmark.class.getResourceAsStream(
                "/" + filename
            )
        ) {
            if (is == null) {
                throw new RuntimeException(
                    "Could not find resource: " + filename
                );
            }
            return mapper.readValue(is, Map.class);
        }
    }

    @SuppressWarnings("unchecked")
    private static void benchmarkAutolink(Map<String, Object> data) {
        List<Map<String, String>> tests = (List<Map<String, String>>) data.get(
            "tests"
        );

        // Warmup old - accumulate to prevent optimization
        for (int i = 0; i < WARMUP_ITERATIONS; i++) {
            for (Map<String, String> test : tests) {
                sink += OLD_AUTOLINK.autoLink(test.get("text")).length();
            }
        }

        // Benchmark old - accumulate lengths
        long startOld = System.nanoTime();
        for (int i = 0; i < ITERATIONS; i++) {
            for (Map<String, String> test : tests) {
                sink += OLD_AUTOLINK.autoLink(test.get("text")).length();
            }
        }
        long elapsedOld = System.nanoTime() - startOld;
        double opsPerSecOld = (ITERATIONS * 1_000_000_000.0) / elapsedOld;

        // Rust FFI version
        try (
            com.sayrer.twitter_text.Autolink rustAutolink =
                new com.sayrer.twitter_text.Autolink()
        ) {
            rustAutolink.setNoFollow(false);

            // Warmup rust - accumulate to prevent optimization
            for (int i = 0; i < WARMUP_ITERATIONS; i++) {
                for (Map<String, String> test : tests) {
                    sink += rustAutolink.autolink(test.get("text")).length();
                }
            }

            // Benchmark rust - accumulate lengths
            long startRust = System.nanoTime();
            for (int i = 0; i < ITERATIONS; i++) {
                for (Map<String, String> test : tests) {
                    sink += rustAutolink.autolink(test.get("text")).length();
                }
            }
            long elapsedRust = System.nanoTime() - startRust;
            double opsPerSecRust = (ITERATIONS * 1_000_000_000.0) / elapsedRust;

            printResults("Autolink", opsPerSecOld, opsPerSecRust);
        }
    }

    @SuppressWarnings("unchecked")
    private static void benchmarkExtract(Map<String, Object> data) {
        Map<String, Object> tests = (Map<String, Object>) data.get("tests");
        List<Map<String, String>> mentions = (List<
            Map<String, String>
        >) tests.get("mentions");
        List<Map<String, String>> urls = (List<Map<String, String>>) tests.get(
            "urls"
        );
        List<Map<String, String>> hashtags = (List<
            Map<String, String>
        >) tests.get("hashtags");
        List<Map<String, String>> cashtags = (List<
            Map<String, String>
        >) tests.get("cashtags");

        // Collect all texts (like JS/Rust benchmark)
        java.util.ArrayList<String> allTexts = new java.util.ArrayList<>();
        for (Map<String, String> test : mentions) {
            allTexts.add(test.get("text"));
        }
        for (Map<String, String> test : urls) {
            allTexts.add(test.get("text"));
        }
        for (Map<String, String> test : hashtags) {
            allTexts.add(test.get("text"));
        }
        for (Map<String, String> test : cashtags) {
            allTexts.add(test.get("text"));
        }

        // Warmup old - call all 4 extract functions for each text
        for (int i = 0; i < WARMUP_ITERATIONS; i++) {
            for (String text : allTexts) {
                sink += OLD_EXTRACTOR.extractMentionedScreennames(text).size();
                sink += OLD_EXTRACTOR.extractURLs(text).size();
                sink += OLD_EXTRACTOR.extractHashtags(text).size();
                sink += OLD_EXTRACTOR.extractCashtags(text).size();
            }
        }

        // Benchmark old
        long startOld = System.nanoTime();
        for (int i = 0; i < ITERATIONS; i++) {
            for (String text : allTexts) {
                sink += OLD_EXTRACTOR.extractMentionedScreennames(text).size();
                sink += OLD_EXTRACTOR.extractURLs(text).size();
                sink += OLD_EXTRACTOR.extractHashtags(text).size();
                sink += OLD_EXTRACTOR.extractCashtags(text).size();
            }
        }
        long elapsedOld = System.nanoTime() - startOld;
        double opsPerSecOld = (ITERATIONS * 1_000_000_000.0) / elapsedOld;

        // Rust FFI version
        try (
            com.sayrer.twitter_text.Extractor rustExtractor =
                new com.sayrer.twitter_text.Extractor()
        ) {
            // Warmup rust - call all 4 extract functions for each text
            for (int i = 0; i < WARMUP_ITERATIONS; i++) {
                for (String text : allTexts) {
                    sink += rustExtractor.extractMentionedScreennames(
                        text
                    ).length;
                    sink += rustExtractor.extractUrls(text).length;
                    sink += rustExtractor.extractHashtags(text).length;
                    sink += rustExtractor.extractCashtags(text).length;
                }
            }

            // Benchmark rust
            long startRust = System.nanoTime();
            for (int i = 0; i < ITERATIONS; i++) {
                for (String text : allTexts) {
                    sink += rustExtractor.extractMentionedScreennames(
                        text
                    ).length;
                    sink += rustExtractor.extractUrls(text).length;
                    sink += rustExtractor.extractHashtags(text).length;
                    sink += rustExtractor.extractCashtags(text).length;
                }
            }
            long elapsedRust = System.nanoTime() - startRust;
            double opsPerSecRust = (ITERATIONS * 1_000_000_000.0) / elapsedRust;

            printResults("Extract", opsPerSecOld, opsPerSecRust);
        }
    }

    @SuppressWarnings("unchecked")
    private static void benchmarkValidateTweet(Map<String, Object> data) {
        Map<String, Object> tests = (Map<String, Object>) data.get("tests");
        List<Map<String, Object>> tweets = (List<
            Map<String, Object>
        >) tests.get("tweets");

        // Warmup old - accumulate valid count
        for (int i = 0; i < WARMUP_ITERATIONS; i++) {
            for (Map<String, Object> test : tweets) {
                sink += OLD_VALIDATOR.isValidTweet((String) test.get("text"))
                    ? 1
                    : 0;
            }
        }

        // Benchmark old - accumulate valid count
        long startOld = System.nanoTime();
        for (int i = 0; i < ITERATIONS; i++) {
            for (Map<String, Object> test : tweets) {
                sink += OLD_VALIDATOR.isValidTweet((String) test.get("text"))
                    ? 1
                    : 0;
            }
        }
        long elapsedOld = System.nanoTime() - startOld;
        double opsPerSecOld = (ITERATIONS * 1_000_000_000.0) / elapsedOld;

        // Rust FFI version
        try (
            com.sayrer.twitter_text.Validator rustValidator =
                com.sayrer.twitter_text.Validator.create()
        ) {
            // Warmup rust - accumulate valid count
            for (int i = 0; i < WARMUP_ITERATIONS; i++) {
                for (Map<String, Object> test : tweets) {
                    sink += rustValidator.isValidTweet(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
            }

            // Benchmark rust - accumulate valid count
            long startRust = System.nanoTime();
            for (int i = 0; i < ITERATIONS; i++) {
                for (Map<String, Object> test : tweets) {
                    sink += rustValidator.isValidTweet(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
            }
            long elapsedRust = System.nanoTime() - startRust;
            double opsPerSecRust = (ITERATIONS * 1_000_000_000.0) / elapsedRust;

            printResults("Validate Tweet", opsPerSecOld, opsPerSecRust);
        }
    }

    @SuppressWarnings("unchecked")
    private static void benchmarkValidateAll(Map<String, Object> data) {
        Map<String, Object> tests = (Map<String, Object>) data.get("tests");
        List<Map<String, Object>> tweets = (List<
            Map<String, Object>
        >) tests.get("tweets");
        List<Map<String, Object>> usernames = (List<
            Map<String, Object>
        >) tests.get("usernames");
        List<Map<String, Object>> hashtags = (List<
            Map<String, Object>
        >) tests.get("hashtags");
        List<Map<String, Object>> urls = (List<Map<String, Object>>) tests.get(
            "urls"
        );

        // Rust FFI version only - old Java doesn't have isValidUsername/Hashtag/Url
        try (
            com.sayrer.twitter_text.Validator rustValidator =
                com.sayrer.twitter_text.Validator.create()
        ) {
            // Warmup rust - all 4 validation types
            for (int i = 0; i < WARMUP_ITERATIONS; i++) {
                for (Map<String, Object> test : tweets) {
                    sink += rustValidator.isValidTweet(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
                for (Map<String, Object> test : usernames) {
                    sink += rustValidator.isValidUsername(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
                for (Map<String, Object> test : hashtags) {
                    sink += rustValidator.isValidHashtag(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
                for (Map<String, Object> test : urls) {
                    sink += rustValidator.isValidUrl((String) test.get("text"))
                        ? 1
                        : 0;
                }
            }

            // Benchmark rust - all 4 validation types
            long startRust = System.nanoTime();
            for (int i = 0; i < ITERATIONS; i++) {
                for (Map<String, Object> test : tweets) {
                    sink += rustValidator.isValidTweet(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
                for (Map<String, Object> test : usernames) {
                    sink += rustValidator.isValidUsername(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
                for (Map<String, Object> test : hashtags) {
                    sink += rustValidator.isValidHashtag(
                            (String) test.get("text")
                        )
                        ? 1
                        : 0;
                }
                for (Map<String, Object> test : urls) {
                    sink += rustValidator.isValidUrl((String) test.get("text"))
                        ? 1
                        : 0;
                }
            }
            long elapsedRust = System.nanoTime() - startRust;
            double opsPerSecRust = (ITERATIONS * 1_000_000_000.0) / elapsedRust;

            System.out.printf("%nValidate All (%d iterations):%n", ITERATIONS);
            System.out.printf("  Rust FFI:  %,.0f ops/sec%n", opsPerSecRust);
        }
    }

    @SuppressWarnings("unchecked")
    private static void benchmarkParse(Map<String, Object> data) {
        List<Map<String, Object>> tests = (List<Map<String, Object>>) data.get(
            "tests"
        );

        // Warmup old - accumulate weighted lengths
        for (int i = 0; i < WARMUP_ITERATIONS; i++) {
            for (Map<String, Object> test : tests) {
                sink += com.twitter.twittertext.TwitterTextParser.parseTweet(
                    (String) test.get("text")
                ).weightedLength;
            }
        }

        // Benchmark old - accumulate weighted lengths
        long startOld = System.nanoTime();
        for (int i = 0; i < ITERATIONS; i++) {
            for (Map<String, Object> test : tests) {
                sink += com.twitter.twittertext.TwitterTextParser.parseTweet(
                    (String) test.get("text")
                ).weightedLength;
            }
        }
        long elapsedOld = System.nanoTime() - startOld;
        double opsPerSecOld = (ITERATIONS * 1_000_000_000.0) / elapsedOld;

        // Rust FFI version
        // Warmup rust - accumulate weighted lengths
        for (int i = 0; i < WARMUP_ITERATIONS; i++) {
            for (Map<String, Object> test : tests) {
                sink += com.sayrer.twitter_text.TwitterTextParser.parseTweet(
                    (String) test.get("text")
                ).weightedLength;
            }
        }

        // Benchmark rust - accumulate weighted lengths
        long startRust = System.nanoTime();
        for (int i = 0; i < ITERATIONS; i++) {
            for (Map<String, Object> test : tests) {
                sink += com.sayrer.twitter_text.TwitterTextParser.parseTweet(
                    (String) test.get("text")
                ).weightedLength;
            }
        }
        long elapsedRust = System.nanoTime() - startRust;
        double opsPerSecRust = (ITERATIONS * 1_000_000_000.0) / elapsedRust;

        printResults("Parse Tweet", opsPerSecOld, opsPerSecRust);
    }

    private static void printResults(
        String operation,
        double oldOpsPerSec,
        double rustOpsPerSec
    ) {
        double speedup = rustOpsPerSec / oldOpsPerSec;
        String label = speedup >= 1 ? "faster" : "slower";
        double ratio = speedup >= 1 ? speedup : (1 / speedup);

        System.out.printf("%n%s (%d iterations):%n", operation, ITERATIONS);
        System.out.printf("  Old Java:  %,.0f ops/sec%n", oldOpsPerSec);
        System.out.printf("  Rust FFI:  %,.0f ops/sec%n", rustOpsPerSec);
        System.out.printf("  Result:    %.1fx %s%n", ratio, label);
    }
}
